mesonbuild/interpreter.py

```diff
@@ -2422,14 +2422,14 @@ def add_build_def_file(self, f):
                 return
             f = os.path.normpath(f.relative_name())
         elif os.path.isfile(f) and not f.startswith('/dev'):
-            srcdir = self.environment.get_source_dir()
-            builddir = self.environment.get_build_dir()
-            f = os.path.normpath(f)
-            rel_path = mesonlib.relpath(f, start=srcdir)
-            if not rel_path.startswith('..'):
-                f = rel_path
-            elif not mesonlib.relpath(f, start=builddir).startswith('..'):
+            srcdir = Path(self.environment.get_source_dir())
+            builddir = Path(self.environment.get_build_dir())
+            f = Path(f).resolve()
+            if builddir in f.parents:
                 return
+            if srcdir in f.parents:
+                f = f.relative_to(srcdir)
+            f = os.path.normpath(f.as_posix())
```

Comment:
Why do we need `normpath()`? Didn't we already call `resolve()` which is a superset of that?

Reply:
because pathlib and normpath does not normalize the same way, so I was getting duplicates in the build_def_files list. IIRC that was causing CI of this PR to fail on Windows.

Reply:
See the doc:
- as_posix: Return a string representation of the path with forward slashes (/)
- normpath: On Windows, it converts forward slashes to backward slashes

Reply:
Ah, but then why use `f.as_posix()` at all? `str(f)` will always use the correct path separator.

Reply:
ah, didn't though about `str(f)`, I updated my patch, let's see if that pass CI.